pkg/controller/certificates/trigger/trigger_controller.go

```diff
@@ -199,27 +199,29 @@ func (c *controller) ProcessItem(ctx context.Context, key string) error {
 	return nil
 }
 
-// shouldBackoffReissuingOnFailure tells us if we should back off
-// re-issuing for this certificate and for how much time.
+// shouldBackoffReissuingOnFailure tells us if we should back-off re-issuing for
+// an hour or not. Notably, it returns no back-off when the certificate doesn't
+// match the "next" certificate (since a mismatch means that this certificate
+// gets re-issued immediately).
 //
-// The request can be left nil, in which case no back off is required.
-// Additionally, if the certificate doesn't match the request, no back
-// off is required because it means the request will have to be re-issued.
+// Note that the request can be left nil: in that case, the returned back-off
+// will be 0 since it means the CR must be created immediately.
 func shouldBackoffReissuingOnFailure(log logr.Logger, c clock.Clock, crt *cmapi.Certificate, nextCR *cmapi.CertificateRequest) (backoff bool, delay time.Duration) {
 	if crt.Status.LastFailureTime == nil {
 		return false, 0
 	}
 
 	// We want to immediately trigger a re-issuance when the certificate
-	// changes, even when it is failing. In order to detect a "change", we
-	// compare the "next" CR with the certificate spec and reissue and there is
-	// a mismatch.
+	// changes. In order to detect a "change", we compare the "next" CR with the
+	// certificate spec and reissue and there is a mismatch.
 	//
 	// Note that the "next" CR is the only CR that matters when looking at
 	// whether the certificate still matches its CR. The "current" CR matches
 	// the previous spec of the certificate, so we don't want to be looking at
 	// the current CR.
-	if nextCR != nil {
+	if nextCR == nil {
```

Comment:
Question: does it actually make sense backing off if there is no next CertificateRequest? Perhaps we should return false here? Just thinking about edge cases where someone deleted the CertificateRequest whilst debugging and now need to wait an hour for changes to Certificate taking effect?

Reply:
Good question, I had to think twice before answering, it requires me a ton of effort in order to keep all this machinery in my brain...

I want to say "no", there does not seem to be any reason to continue this function further if the next CR does not exist.

After a bit more thinking, I remembered that the certificate's `durationSinceFailure` may have been set for other reasons that may come earlier than the creation of the "next" certificate request: for example, when the key manager sees that the existing private key does not have the correct key size, the certificate will have `durationSinceFailure` and at the same time the above `nextCR` will be nil.

Does it make sense?

Reply:
Thank you, it makes sense!

Although now I am thinking whether it could ever happen that the next CR is nil as it was never issued due to Certificate'
s misconfiguraton, we are in failure mode and the user updates the Certificate to fix the failure and we should re-issue immediately 😆 
I would like a simulator of all the possible paths 😄